rosserial_chibios/src/rosserial_chibios/make_libraries.py

```diff
@@ -0,0 +1,94 @@
+#    from this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+THIS_PACKAGE = "rosserial_chibios"
+
+__usage__ = """
+make_libraries.py generates the ChibiOS rosserial library files. It
+requires the location of your ChibiOS project folder.
+
+rosrun rosserial_chibios make_libraries.py <output_path>
+"""
+
+import rospkg
+
+import rosserial_client
+from rosserial_client.make_library import *
+
+# for copying files
+import shutil
+
+ROS_TO_EMBEDDED_TYPES = {
+    'bool'    :   ('bool',              1, PrimitiveDataType, []),
+    'byte'    :   ('int8_t',            1, PrimitiveDataType, []),
+    'int8'    :   ('int8_t',            1, PrimitiveDataType, []),
+    'char'    :   ('uint8_t',           1, PrimitiveDataType, []),
+    'uint8'   :   ('uint8_t',           1, PrimitiveDataType, []),
+    'int16'   :   ('int16_t',           2, PrimitiveDataType, []),
+    'uint16'  :   ('uint16_t',          2, PrimitiveDataType, []),
+    'int32'   :   ('int32_t',           4, PrimitiveDataType, []),
+    'uint32'  :   ('uint32_t',          4, PrimitiveDataType, []),
+    'int64'   :   ('int64_t',           8, PrimitiveDataType, []),
+    'uint64'  :   ('uint64_t',          8, PrimitiveDataType, []),
+    'float32' :   ('float',             4, PrimitiveDataType, []),
+    'float64' :   ('double',            8, PrimitiveDataType, []),
+    'time'    :   ('ros::Time',         8, TimeDataType, ['ros/time']),
+    'duration':   ('ros::Duration',     8, TimeDataType, ['ros/duration']),
+    'string'  :   ('char*',             0, StringDataType, []),
+    'Header'  :   ('std_msgs::Header',  0, MessageDataType, ['std_msgs/Header'])
+}
+
+# Enforce correct inputs
+if (len(sys.argv) < 2):
+    print(__usage__)
+    exit(1)
+
+# Sanitize output path
+path = sys.argv[1]
+if path[-1] == "/":
+    path = path[0:-1]
+print("\nExporting to %s" % path)
+
+rospack = rospkg.RosPack()
+
+# copy ros_lib stuff in
+rosserial_chibios_dir = rospack.get_path(THIS_PACKAGE)
+shutil.copytree(rosserial_chibios_dir+"/src/ros_lib", path+"/ros_lib")
+rosserial_client_copy_files(rospack, path+"/ros_lib/")
+
+# generate messages
+rosserial_generate(rospack, path+"/ros_lib", ROS_TO_EMBEDDED_TYPES)
+
```

Comment:
I know this is just a c&p of similar other files in this repo, but please lose the trailing line and add spaces around the `+` operators above (or better yet, switch them to be `os.path.join`).

Reply:
So be it.

`os.path.join` strips trailing slashes from the path; which breaks further generation of `ros_lib`.

There'll be a PR to migrate everything to `os.path.join` in a minute.

Reply:
Ha, appreciated.